# YouTubeTutorials


This repository holds scripts for demo tutorials on my youtube [channel](https://www.youtube.com/channel/UCnXs-Nq1dzMZQOKUHKW3rdw).

| Script | Youtube Video |
| --- | --- |
| [dataManipulation.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/dataManipulation.R) | [How to manipulate gene expression data from NCBI GEO in R using dplyr](https://www.youtube.com/watch?v=4CkRXGWmAbU&t=1s) |
| [visualize_ggplot2.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/visualize_ggplot2.R) | [Visualize gene expression data in R using ggplot2](https://www.youtube.com/watch?v=RukuTtiY4Sg&t=11s) |
| [dataformats.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/dataformats.R) | [Reading single-cell data in R: H5AD, loom, MEX, AnnData formats](https://youtu.be/3xcTpqQzUwQ) |
| [singleCell_standard_workflow.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_standard_workflow.R) | [How to analyze single-cell RNA-Seq data in R](https://www.youtube.com/watch?v=5HBzgsz8qyk) |
| [runDESeq2.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/runDESeq2.R) , [getData.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/getData.R) | [DESeq2 workflow tutorial](https://youtu.be/OzNzO8qwwp0) |
| [singleCell_integration.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_integration.R) | [Integrate single-cell RNA-Seq datasets in R using Seurat (CCA)](https://youtu.be/HrbeaEJqKcY) |
| [singleCell_doublets.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_doublets.R)  |  [DoubletFinder: Detect doublets in single-cell RNA-Seq data in R](https://youtu.be/NqvAS4HgmrE) |
| [microrray_RMA_normalize.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/microrray_RMA_normalize.R) | [How to read and normalize microarray data in R - RMA normalization](https://youtu.be/TtlvC67pdOM)|
| [singleCell_integrate_harmony.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_integrate_harmony.R) | [Integrate single-cell RNA-Seq data in R using Harmony](https://youtu.be/zEuqhiu341I) |
| [singleCell_CI_markers.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_CI_markers.R) | [Find markers and cluster identification in single-cell RNA-Seq using Seurat](https://youtu.be/1i6T9hpvwg0) |
| [singleCell_pseudoBulk.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/singleCell_pseudoBulk.R) | [Pseudo-bulk analysis for single-cell RNA-Seq data](https://youtu.be/04gB2owLKus) |
| [convert_geneID_geneSymbols.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/convert_geneID_geneSymbols.R) | [3 ways to convert Ensembl IDs to gene symbols](https://youtu.be/cWe359VnfaY) |
| [RNASeqpipeline.sh](https://github.com/kpatel427/YouTubeTutorials/blob/main/RNASeqpipeline.sh) | [Setup RNA-Seq Pipeline from scratch: fastq (reads) to counts](https://youtu.be/lG11JjovJHE)|
| [TI_monocle3.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/TI_monocle3.R) | [Single-cell Trajectory analysis using Monocle3 and Seurat](https://youtu.be/iq4T_uzMFcY) |
| [WGCNA.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/WGCNA.R) | [Weighted Gene Co-expression Network Analysis (WGCNA) Step-by-step Tutorial - Part 1](https://youtu.be/gYE59uEMXT4) & [Weighted Gene Co-expression Network Analysis (WGCNA) Step-by-step Tutorial - Part 2](https://youtu.be/mzXIxjPr_Mc) |
| [variant_calling.sh](https://github.com/kpatel427/YouTubeTutorials/blob/main/variant_calling.sh) | [WGS Variant Calling: Variant calling with GATK - Part 1](https://youtu.be/iHkiQvxyr5c) |
| [DESeq2ErrorFix.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/DESeq2ErrorFix.R) | [DESeq2 Error Fix: DESeqDataSetFromMatrix ncol(countData) == nrow(colData) is not TRUE)](https://youtu.be/nks7ibkBud8) |
| [variant_filtering_annotation.sh](https://github.com/kpatel427/YouTubeTutorials/blob/main/variant_filtering_annotation.sh) | [WGS Variant Calling: Variant Filtering and Annotation - Part 2](https://youtu.be/XZ8scaScfjw) |
| [installPackages.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/installPackages.R) | [How to install packages in R? What is CRAN? What is Bioconductor?](https://youtu.be/GM6MCBkVNtQ) |
| [tcga_data_download.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/tcga_data_download.R) | [Download data from GDC Portal using TCGAbiolinks R Package](https://youtu.be/UWXv9dUpxNE) |
| [KM_survival.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/KM_survival.R) | [Survival analysis with TCGA data in R - Create Kaplan-Meier Curves](https://youtu.be/DnygUTAZFmM) |
| [annotateSingleR.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/annotateSingleR.R) |[Automatic cell-annotation for single-cell RNA-Seq data: A detailed SingleR tutorial (PART 1)](https://youtu.be/7RuPGaWcY0Y) |
| [annotateSingleR_multipleRefs.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/annotateSingleR_multipleRefs.R) |[Automatic cell-annotation for single-cell RNA-Seq data: A detailed SingleR tutorial (PART 2)](https://youtu.be/cZsF0DfeMGg) |
| [setSeed.R](https://github.com/kpatel427/YouTubeTutorials/blob/main/setSeed.R) | [Understanding set.seed in R: Ensuring Reproducibility in Data Analysis](https://youtu.be/M5Yggz3AV1g) |
| [conda and renv scripts](https://github.com/kpatel427/YouTubeTutorials/tree/main/files/conda_virtualEnv) | [Demystifying Conda (Anaconda, Miniconda and Bioconda) and Virtual Environments](https://youtu.be/2bXGm0ZnJ38) |
